pandas/indexes/multi.py

```diff
@@ -656,9 +659,58 @@ def _has_complex_internals(self):
         return True
 
     @cache_readonly
+    def is_monotonic(self):
+
+        # TODO
+        # this is unfortunate we end up tupelizing
+        # just to determine monotonicity :<
+
+        # fast-path
+        if not self.levels[0].is_monotonic:
+            return False
+
```

Comment:
@jreback are you trying to do this just from `self.labels` and `self.levels`, or are you okay with materializing the level values arrays?  If the latter, I think you can do this with `np.lexsort`:

```
# lexsort_test.py

import numpy as np
import pandas as pd

sorted_index = pd.MultiIndex.from_product([[0, 1, 2], ['a', 'b', 'c']])
unsorted_first = pd.MultiIndex.from_product([[0, 2, 1], ['a', 'b', 'c']])
unsorted_second = pd.MultiIndex.from_product([[0, 1, 2], ['a', 'c', 'b']])
duplicates = pd.MultiIndex.from_product([[0, 1, 2], ['b', 'b', 'b']])


def level_values(mi, num):
    return np.take(mi.levels[num], mi.labels[num])


def int64_is_monotonic(a):
    # This is needlessly expensive in the non-monotonic case. I assume this is
    # already implemented efficiently somewhere else in pandas.
    return (np.diff(a) >= 0).all()


def multi_index_is_monotonic(mi):
    # reversed() because lexsort() wants the most significant key last.
    values = [level_values(mi, i) for i in reversed(range(len(mi.levels)))]
    sort_order = np.lexsort(values)
    return int64_is_monotonic(sort_order)


print("Sorted: ", multi_index_is_monotonic(sorted_index))
print("Sorted With Duplicates:", multi_index_is_monotonic(duplicates))
print("Unsorted First:", multi_index_is_monotonic(unsorted_first))
print("Unsorted Second:", multi_index_is_monotonic(unsorted_second))
```
```
$ python lexsort_test.py
Sorted:  True
Unsorted First: False
Unsorted Second: False
Sorted With Duplicates: True
```

One caveat here is that I'm not 100% sure that `lexsort` is guaranteed to be stable, which I think you'd need to avoid false negatives in the case where there are duplicates.  The docs for `lexsort` itself don't seem to mention stability, but the `See Also` from `argsort` says: `lexsort : Indirect stable sort with multiple keys.`

One cv